AWS Lambda Executor

o-nikolas opened the pull request:
### Overview

Introducing the new **AWS Lambda Executor**!

As with our previous executors this is an initial release with most functionality in place. However, there are surely future upgrades and iterations to come, so stay tuned!

This changeset includes:
- The Lambda executor code itself
- Tests for the executor
- Docs for the executor

### Testing
The unit tests cover 90+% line coverage (ignore the hook coverage, the results were filtered on lambda and that hook is prexisting code not delivered in this PR) and the executor was UAT tested manually in many scenarios.

![Screenshot from 2025-05-12 16-51-49](https://github.com/user-attachments/assets/54bf0dea-b880-48b1-ac8e-f1e9fa298cb9)

### Review Notes
This is a large PR, which is not entirely ideal. However, it can be difficult to release portions of a component like this and I've scoped all the code here to the minimum required for folks to begin using this executor. There are other refactorings that were left out and will be contributed at a later time to minimize the confusion.

There are clear chunks this can be reviewed in (docs, provider.yaml, tests, and core logic) to split it up. 

Ultimately, (almost) all the code is scoped to the Amazon provider package and it is a net new component, so there is a very limited blast radius. Little to no existing user workflows or code should be affected.


<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

o-nikolas committed:
AWS Lambda Executor

This changeset includes:
- The Lambda executor code itself
- Tests for the executor
- Docs for the executor

o-nikolas committed:
Small fixes and tweaks from PR tests

PR tests turned up some things that local testing/pre-commit did not
catch. Fix those up.

o-nikolas committed:
More changes from PR checks

o-nikolas committed:
Merge branch 'main' into onikolas/lambda_exec

o-nikolas committed:
Merge branch 'main' into onikolas/lambda_exec

vincbeck commented on providers/amazon/provider.yaml:
```diff
@@ -1003,6 +1003,82 @@ config:
+        default: "3"
+      queue_url:
+        description: |
+          The name of the SQS queue to use for the Lambda executor. Required.
```
name or URL?

o-nikolas replied:
Good catch! Should be URL, will fix

vincbeck commented on providers/amazon/provider.yaml:
```diff
@@ -1003,6 +1003,82 @@ config:
+        default: ~
+      dead_letter_queue_url:
+        description: |
+          The name of the SQS dead letter queue to use for the Lambda function. The Lambda executor
```
name or URL?

o-nikolas replied:
Same as above

vincbeck commented on providers/amazon/docs/executors/lambda-executor.rst:
```diff
@@ -0,0 +1,364 @@
+AWS Lambda Executor
+===================
+
+This is an Airflow executor powered by AWS Lambda. Each task that Airflow schedules
```
Should we mention that Lambda has a hard limit of 15 minutes, thus, any task taking more than 15 minutes will fail?

o-nikolas replied:
Sure, to me I think that's the hallmark thing everyone knows about Lambda, but you're right that it's worth mentioning again, just to cover our bases.

vincbeck commented on providers/amazon/src/airflow/providers/amazon/aws/executors/aws_lambda/docker/Dockerfile:
```diff
@@ -0,0 +1,106 @@
+# NOTE: If you change the below line, specifically removing the amazon extra, please ensure boto3 is
+# installed via another method. Boto3 is required for the Lambda executor to function properly.
+# hadolint ignore=SC2102
+RUN pip install --no-cache-dir --upgrade pip && pip install --no-cache-dir apache-airflow[amazon,postgres]==2.10.5
```
Not Airflow 3?

o-nikolas replied:
Ooops, that's just from testing, good catch. I intend to leave it at latest version.

o-nikolas committed:
Feedback from PR

vincbeck commented on providers/amazon/provider.yaml:
```diff
@@ -1003,6 +1003,82 @@ config:
+      conn_id:
+        description: |
+          The Airflow connection (i.e. credentials) used by the Lambda executor to make API calls.
+        version_added: "9.7.0"
```
`9.8.0rc1` is out so it will most likely be released in `9.9.0`

```suggestion
        version_added: "9.9.0"
```

o-nikolas replied:
Ya fair, this thing is annoying since you're chasing it until it's merged, will update.

o-nikolas replied:
done

vincbeck commented on providers/amazon/src/airflow/providers/amazon/aws/executors/aws_lambda/lambda_executor.py:
```diff
@@ -0,0 +1,488 @@
+    MAX_INVOKE_ATTEMPTS = conf.get(
+        CONFIG_GROUP_NAME,
+        AllLambdaConfigKeys.MAX_INVOKE_ATTEMPTS,
+        fallback=3,
```
It applies to all `conf.get` calls, the `fallback` should be necessary. The default is already defined in config

o-nikolas replied:
This one is weird, I thought so too, but my testing has shown that this does not work as expected? Maybe it's something on my end?

vincbeck replied:
Did you test it after you run `pre-commits`? `pre-commits` generate `get_provider_info.py` which are used to get configuration for providers. So you must run pre-commits to generate these file and then after these configuration are getting picked-up.

Another idea would be, `@providers_configuration_loaded` is missing. This decorator loads the configuration from providers. Since the executor is instantiated in the scheduler component, we do not load configuration from providers in scheduler?

o-nikolas replied:
Interesting, I'll go have a look at these two options!

vincbeck commented on providers/amazon/src/airflow/providers/amazon/aws/executors/aws_lambda/lambda_executor.py:
```diff
@@ -0,0 +1,488 @@
+        self.log.info("Loading Connections")
+        aws_conn_id = conf.get(CONFIG_GROUP_NAME, AllLambdaConfigKeys.AWS_CONN_ID, fallback="aws_default")
+        region_name = conf.get(CONFIG_GROUP_NAME, AllLambdaConfigKeys.REGION_NAME, fallback=None)
+        self.sqs_client = SqsHook(aws_conn_id=aws_conn_id, region_name=region_name).conn
+        self.lambda_client = LambdaHook(aws_conn_id=aws_conn_id, region_name=region_name).conn
```
Do not you think we should cache them?

o-nikolas replied:
What do you mean? The same connection is used unless we detect an issue with it (like expired tokens) and then this function will be called again only in that case.

vincbeck replied:
I meant the hooks, do you think it is worth caching them with `@cached_property`. As far as I know, hook creation can be heavy?

o-nikolas replied:
We actually need new hooks to get new connections (since the hooks cache their connections), which is the goal here to get a brand new connection since the last one is unhealthy. And again, we only call it when the connection is unhealthy which should be rare, and if you have an unhealthy connection your executor will be completely broken and and non-functioning, so I'm sure the user will be happy to pay a small time/cpu price to fix up everything!

vincbeck replied:
Indeed, sounds very valid to me :) Thanks for clarifying

o-nikolas committed:
Newer lambda images use dnf not yum

eladkal commented on providers/amazon/src/airflow/providers/amazon/aws/executors/aws_lambda/docker/Dockerfile:
```diff
@@ -0,0 +1,106 @@
+# under the License.
+
+# Use the official AWS Lambda Python base image.
+FROM public.ecr.aws/lambda/python:3.12
```
Can you clarify about this?
what happens if I run Airflow with Python 3.11 and want to use the lambda executor?

o-nikolas replied:
Then the user should update the Dockerfile to reflect the version they would like to use. This Dockerfile is only an example and cannot cover all user stories. It's completely up to the user to take this example and use/modify it to suit their needs.

I'll add a comment above that line suggesting the user to update it if they wish to use another python version.

o-nikolas committed:
Update provider version

o-nikolas committed:
Remove redundant config fallbacks

o-nikolas committed:
Move max invoke attempts to be set during construction

o-nikolas committed:
Add comment regarding python version to Dockerfile

gopidesupavan commented on providers/amazon/src/airflow/providers/amazon/aws/executors/aws_lambda/lambda_executor.py:
```diff
@@ -0,0 +1,484 @@
+            )
+
+            def _check_queue(queue_url):
+                sqs_get_queue_attrs_response = self.sqs_client.get_queue_attributes(
```
It would be better to move this inside the try block, so that we catch exception.

o-nikolas replied:
Yupp, fair! This all runs inside a larger try/except so it is ultimately caught, but that is a bit strange. Let me refactor.

gopidesupavan commented on providers/amazon/src/airflow/providers/amazon/aws/executors/aws_lambda/lambda_executor.py:
```diff
@@ -0,0 +1,484 @@
+                        queue_url,
+                        approx_num_msgs,
+                    )
+                except Exception:
```
Logging exception will give user better debugging option, what exactly the error thrown, in this case mostly we get permission errors/ or queue itself is wrong  or not exists :)

o-nikolas replied:
Yeah that's fair, the context of the original exception is getting dropped here which would be helpful to have for the user to fix the issue. Let me log it

o-nikolas replied:
I simplified the try/catches, removing all but the outer try/catch. It will except all exceptions, log the exception with the context that the lambda executor is not healthy and the issue needs rectifying

gopidesupavan replied:
cool :)

o-nikolas committed:
Simplify function/queue health checks

o-nikolas committed:
Remove executor_config

Passing through clientContext to the lambda function is only supported
on synchronous executions.

o-nikolas committed:
Merge branch 'main' into onikolas/lambda_exec

eladkal commented on providers/amazon/src/airflow/providers/amazon/aws/executors/aws_lambda/lambda_executor.py:
This path is not according to the convention. We shouldn't have `aws_` prefix.

o-nikolas replied:
`lambda` is a reserved keyword in Python, so many of the places I needed to prefix it with something. I prefixed in all cases for consistency. It is unfortunate though

eladkal replied:
got it! thanks
